pkg/processor/procbuiltin/unwrap.go

```diff
@@ -0,0 +1,310 @@
+	data := rec.Payload.After
+	var structData record.StructuredData
+	switch d := data.(type) {
+	case record.RawData:
+		// todo: take this section out, after platform team support ordering processors
+		// unmarshal raw data to structured
+		err := json.Unmarshal(data.Bytes(), &structData)
+		if err != nil {
+			return record.Record{}, cerrors.Errorf("failed to unmarshal raw data as JSON: %w", unwrapPayloadProcType, err)
+		}
+	case record.StructuredData:
+		structData = d
+	default:
+		return record.Record{}, cerrors.Errorf("unexpected data type %T", unwrapPayloadProcType, data)
+	}
+
+	// get payload
+	var structPayload record.StructuredData
+	payload := structData["payload"] // the payload has the debezium record
+	if p, ok := payload.(map[string]interface{}); ok {
+		structPayload = p
+	} else {
+		return record.Record{}, cerrors.Errorf("payload doesn't contain a record")
+	}
+
+	// check fields under payload
+	err := d.validateRecord(structPayload)
+	if err != nil {
+		return record.Record{}, err
+	}
+
+	var before record.StructuredData
+	if b, ok := structPayload[DebeziumFieldBefore].(map[string]any); ok {
+		before = b
+	}
+
+	var after record.StructuredData
+	if a, ok := structPayload[DebeziumFieldAfter].(map[string]any); ok {
+		after = a
+	}
+
+	var op string
+	if str, ok := structPayload[DebeziumFieldOp].(string); ok {
+		op = str
+	} else {
+		return record.Record{}, cerrors.Errorf("%s operation is not a string", op)
+	}
+
+	operation, err := d.convertOperation(op)
+	if err != nil {
+		return record.Record{}, cerrors.Errorf("error unwrapping operation: %w", err)
+	}
+
+	metadata, err := d.unwrapMetadata(rec, structPayload)
+	if err != nil {
+		return record.Record{}, cerrors.Errorf("error unwrapping metadata: %w", err)
+	}
+
+	return record.Record{
+		Key:       rec.Key,
+		Position:  rec.Position,
+		Operation: operation,
+		Payload: record.Change{
+			Before: before,
+			After:  after,
+		},
+		Metadata: metadata,
+	}, nil
+}
+
+func (d *DebeziumUnWrapper) validateRecord(data record.StructuredData) error {
+	var multiErr error
+	if _, ok := data[DebeziumFieldBefore]; !ok {
+		multiErr = multierror.Append(multiErr, cerrors.Errorf("the %q field is missing from debezium payload", DebeziumFieldBefore))
+	}
+	if _, ok := data[DebeziumFieldAfter]; !ok {
+		multiErr = multierror.Append(multiErr, cerrors.Errorf("the %q field is missing from debezium payload", DebeziumFieldAfter))
+	}
+	if _, ok := data[DebeziumFieldSource]; !ok {
+		multiErr = multierror.Append(multiErr, cerrors.Errorf("the %q field is missing from debezium payload", DebeziumFieldSource))
+	}
+	if _, ok := data[DebeziumFieldOp]; !ok {
+		multiErr = multierror.Append(multiErr, cerrors.Errorf("the %q field is missing from debezium payload", DebeziumFieldOp))
+	}
+	// ts_ms and transaction can be empty
+	return multiErr
+}
+
+func (d *DebeziumUnWrapper) unwrapMetadata(rec record.Record, payload record.StructuredData) (record.Metadata, error) {
```

Comment:
The record carries the payload as well, no need to pass it separately.
```suggestion
func (d *DebeziumUnWrapper) unwrapMetadata(rec record.Record) (record.Metadata, error) {
```

Reply:
I passed it because the payload could be raw data and I would've needed to parse it again, but now since we expect it to only be structured, then it will work